src/torrent.cpp

```diff
@@ -1138,6 +1138,50 @@ bool is_downloading_state(int const st)
 		if (m_peer_list) m_peer_list->clear();
 	}
 
+	void torrent::set_sequential_range(piece_index_t first_piece, piece_index_t last_piece)
+	{
+		if (!has_picker()) {
+			if (!valid_metadata() || !m_connections_initialized) return;
+			else if (!m_have_all
+				|| settings().get_int(settings_pack::suggest_mode)
+				== settings_pack::suggest_read_cache)
+				need_picker();
+			else return;
+		}
+		m_sequential_download = true;
+		m_picker->set_sequential_range(first_piece, last_piece);
+	}
+
+	void torrent::set_sequential_start(piece_index_t first_piece)
+	{
+		if (!has_picker()) {
+			if (!valid_metadata() || !m_connections_initialized) return;
+			else if (!m_have_all
+				|| settings().get_int(settings_pack::suggest_mode)
+				== settings_pack::suggest_read_cache)
+				need_picker();
+			else return;
+		}
+		// m_torrent_file->files() is file_storage
+		// find the last file in piece `first_piece`
+		auto const f = m_torrent_file->files().map_block(first_piece, m_torrent_file->files().piece_size(first_piece) - 1, 1).back().file_index;
+
+		// find the last piece in file `f`
+		std::int64_t piece_size = m_torrent_file->files().piece_length();
+		auto const last_piece = piece_index_t(
+			std::lldiv(
+			(m_torrent_file->files().file_offset(f)
+			+
+			m_torrent_file->files().file_size(f)
+			-
+			1)
+			,///
```

Comment:
I think this comment should be removed, and this whole expression should probably be formatted in a bit more readable way.

Additionally, I think this function belongs as a member on file_storage. Partly to make it easier for clients to access in general, but also to allow it to be unit tested. maybe something like file_storage::file_first_piece() and file_storage::file_last_piece().

Currently, there is no unit test ensuring that finding the last piece of the file works.

Reply:
You're commenting on outdated commits here.